src/test/scala/chisel3/tests/AsyncResetRegTest.scala

```diff
@@ -0,0 +1,159 @@
+    name match {
+      case "d"     => nextValue = value
+      case "en"    => enable = value > BigInt(0)
+      case "rst"   =>
+        if(value > BigInt(0)) {
+          nextValue = resetValue
+          currentValue = nextValue
+        }
+      case _ =>
+    }
+    println(s"next $nextValue cur $currentValue, en $enable")
+  }
+
+  override def clockChange(transition: Transition, clockName: String): Unit = {
+    if(transition == PositiveEdge && enable) {
+      currentValue = nextValue
+    }
+  }
+
+  override def outputDependencies(outputName: String): Seq[String] = {
+    Seq("rst", "clk", "d")
+  }
+
+  override def setParams(params: Seq[Param]): Unit = {
+    params.foreach {
+      case firrtl.ir.IntParam("RESET_VALUE", value) =>
+        resetValue = value
+      case _ =>
+      // ignore
+    }
+  }
+}
+
+class AsyncResetBlackBoxFactory extends ScalaBlackBoxFactory {
+  override def createInstance(instanceName: String, blackBoxName: String): Option[ScalaBlackBox] = {
+    blackBoxName match {
+      case "AsyncResetReg" =>
+        Some(add(new AsyncResetRegScalaImpl(instanceName)))
+      case _ =>
+        None
+    }
+  }
+}
+
+class SimpleRegIO(val w: Int) extends Bundle{
+  val d = Input(UInt(w.W))
+  val q = Output(UInt(w.W))
+  val en = Input(Bool())
+}
+
+class AsyncResetRegVec(val w: Int, val init: BigInt) extends Module {
+  val io: SimpleRegIO = IO(new SimpleRegIO(w))
+
+  private val async_regs = List.tabulate(w) { idx =>
+    val on = if (init.testBit(idx)) 1 else 0
+    Module(new AsyncResetReg(on))
+  }
+
+  private val q = for ((reg, idx) <- async_regs.zipWithIndex) yield {
+    reg.io.clk := clock
+    reg.io.rst := reset
+    reg.io.d   := io.d(idx)
+    reg.io.en  := io.en
+    reg.suggestName(s"reg_$idx")
+    reg.io.q
+  }
+
+  //endian
+  io.q := q.reverse.map(_.asUInt()).reduce { (a, b) => Cat(a, b) }
+
+  override def desiredName = s"AsyncResetRegVec_w${w}_i$init"
+
+}
+
+class UsesAsyncResetReg extends Module {
+  val io = IO(new Bundle {
+    val in = Input(UInt(AsyncResetReg.bitWidth.W))
+    val out = Output(UInt(AsyncResetReg.bitWidth.W))
+  })
+
+  val reg = Module(new AsyncResetRegVec(AsyncResetReg.bitWidth, init = 3))
+
+  reg.io.d := io.in
+  reg.clock := clock
+  reg.reset := reset
+  reg.io.en := 1.U
+
+  io.out := reg.io.q
+}
+
+class AsyncResetRegTest extends FreeSpec with ChiselScalatestTester {
+
+  "it should work normally with respect to clock" in {
+
+    val manager = new Testers2OptionsManager {
+      treadleOptions = treadleOptions.copy(
+        blackBoxFactories = Seq(new AsyncResetBlackBoxFactory)
+      )
+    }
+    test(new UsesAsyncResetReg, manager) { c =>
+
+      // testers default reset will set initial value to 3
+      c.io.in.poke(7.U)
+      c.io.out.expect(3.U)
+      c.clock.step()
+      c.io.out.expect(7.U)
+
+      // reset immediately is visible with after reset is poked again
+      c.reset.poke(true.B)
+      c.io.out.expect(3.U)
```

Comment:
I'd also test these conditions:
- Register doesn't update when reset is asserted and its clock steps
- Register updates on clock after reset is deasserted (yes, this is technically handled by testers2 infrastructure, but still...)
- Register maintains reset value after reset deassert but before clock step
- With the single-bit AsyncResetReg, maybe instantiate two of them, one resetting to 0 and the other to 1

Reply:
I think I got all these please confirm